### PR TITLE
feat: Allow declaring cfg groups in rust-project.json, to help sharing common cfgs

### LINKED ISSUE #17815: Specifying a set of `cfg`s for given crates (`cfg` groups)
The Linux kernel has ~20k (yes, kilo!) configuration options, and we pass them via `--cfg`s. This means that, in our `rust-project.json` file, we have to put (a subset of) all of them, repeated per crate.

Currently we only have a small number of crates, but our configuration file can already reach 150k lines, ~7.5 MiB with just 6 crates (plus `core` etc. that do not need the `cfg`s). As Rust in the kernel grows, and as we start splitting our `kernel` crate in smaller ones, it can get out of control soon: we may have 10 now, but we can eventually end up with 1000s if Rust is a success. Thus the file size may explode, and I imagine performance and memory usage may not be great either, depending on how it is implemented.

From a quick look at the documentation, it does not seem like one can define a set of "global" `cfg`s to apply to "all" crates. Ideally, one would be able to define "groups" and then apply those groups to some crates before their individual set of `cfg`s.

So, for instance, in our case, we would just define a group like `"kernel": ["CONFIG_...=...", ...],` with essentially all the `cfg`s, and then in each kernel crate that needs them we would have a key like: `"cfg_groups": ["kernel"],`, rather than repeating every time all the `cfg`s in the existing `"cfg"` key.

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
Closes #17815.

### CODE REVIEW

File: crates/project-model/src/project_json.rs [hunk 1]
```diff
@@ -320,8 +345,9 @@ struct CrateData {
     version: Option<semver::Version>,
     deps: Vec<Dep>,
     #[serde(default)]
-    #[serde(with = "cfg_")]
-    cfg: Vec<CfgAtom>,
+    cfg_extends: FxHashSet<String>,
```
Comment (Veykril) [inline crates/project-model/src/project_json.rs:348]:
```suggestion
    cfg_groups: FxHashSet<String>,
```
I think that name fits better though either works I'd say

File: crates/project-model/src/project_json.rs [hunk 2]
```diff
@@ -319,9 +346,11 @@ struct CrateData {
     #[serde(default)]
     version: Option<semver::Version>,
     deps: Vec<Dep>,
+    // Cannot be a hashset because that would mean an inconsistent order.
+    #[serde(default)]
+    cfg_groups: Vec<String>,
```
Comment (davidbarsky) [inline crates/project-model/src/project_json.rs:351]: I think that if you say this is `FxHashSet`, order would be consistent across invocations.

File: crates/project-model/src/project_json.rs [hunk 3]
```diff
@@ -122,6 +122,27 @@ impl ProjectJson {
...
+                        .flat_map(|cfg_extend| {
+                            let cfg_group = data.cfg_groups.get(cfg_extend);
+                            match cfg_group {
+                                Some(cfg_group) => cfg_group.0.iter().cloned(),
+                                None => {
+                                    tracing::error!(
+                                        "Unknown cfg group `{cfg_extend}` in crate `{}`",
+                                        crate_data.display_name.as_deref().unwrap_or("<unknown>"),
+                                    );
+                                    [].iter().cloned()
```
Comment (davidbarsky) [inline crates/project-model/src/project_json.rs:139]: More of a large meta point, but I think it's be a good idea to us to consider using https://github.com/rust-lang/annotate-snippets-rs for rust-analyzer-specific errors.

File: crates/project-model/src/project_json.rs [hunk 4]
```diff
@@ -122,6 +122,27 @@ impl ProjectJson {
                         None => None,
                     };
 
+                    crate_data.cfg_groups.sort_unstable();
+                    crate_data.cfg_groups.dedup();
```
Comment (davidbarsky) [inline crates/project-model/src/project_json.rs:126]: I don't think you'd need this if you're to use `FxHashSet`.

**Review by davidbarsky — COMMENTED** [review]
Not opposed to this change in principle; just had two comments:
- Considering using `FxHashSet`.
- I don't know if the memory usage for rust-analyzer would be better (as mentioned as a motivating example in https://github.com/rust-lang/rust-analyzer/issues/17815 by @ojeda), but I can absolutely imagine that the _file_ size would be better. However, if Rust for Linux would prefer to use the automatic indexing approach I described in https://github.com/rust-lang/rust-analyzer/issues/17816#issuecomment-2284066355, maybe that matters less?

Anyways, I think this is a reasonable addition; I just want to make sure that some details of my understanding are not out date.

**Comment by davidbarsky** [comment]
Gotcha, yeah. We can probably land this this largely as-is and revisit in a year if its carrying its weight.

**Comment by ChayimFriedman2** [comment]
I can make the crates share the cfg in memory too (i.e. not expand them when constructing the crate graph). It is not even hard, but it means when checking cfgs we will have to check for parent groups too, which may have an impact on rust-analyzer performance. I don't really know how to benchmark this, and if it indeed regresses perf, it raises the question of whether it is worth regressing perf for everyone to please the few that have a large amount of shared cfgs.

**Comment by Veykril** [comment]
I wouldn't worry about the memory usage impact for this I think, the main compelling argument to me was the file size reduction. But indeed, if the discovery mechanism solves this problem then I'd rather get rid of this (as in im fine with us landing this temporarily but removing it once its not longer needed by the linux project

File: crates/project-model/src/project_json.rs (hunk 3 above)
Comment (Veykril) [reply crates/project-model/src/project_json.rs:137]: That seems tailored to CLI output, doesn't sound too useful for us

**Comment by Veykril** [comment]
Will merge this for now, if the linux project doesn't need this anymore in the future (due to project discovery etc) let us know then we'll likely remove it again
@bors r+

**Comment by ojeda** [comment]
> Will merge this for now, if the linux project doesn't need this anymore in the future (due to project discovery etc) let us know then we'll likely remove it again @bors r+

Thanks!

**Pull request merged**